design/design-doc-packages-v2.md

```diff
@@ -0,0 +1,1900 @@
+
+  // ImagePullPolicy defines the pull policy for all images used during
+  // Package extraction and when running the Package controller.
+  // https://kubernetes.io/docs/concepts/configuration/overview/#container-images
+  ImagePullPolicy corev1.PullPolicy `json:"imagePullPolicy,omitempty"`
+
+  // ServiceAccount options allow for changes to the ServiceAccount the
+  // Package Manager creates for the Package's controller
+  ServiceAccount *ServiceAccountOptions `json:"serviceAccount,omitempty"`
+}
+
+// CRDList is the full list of CRDs that this package owns and depends on
+type CRDList []metav1.TypeMeta
+
+// ControllerSpec defines the controller that implements the logic for a
+// package, which can come in different flavors.
+type ControllerSpec struct {
+  // ServiceAccount options allow for changes to the ServiceAccount the
+  // Package Manager creates for the Package's controller
+  ServiceAccount *ServiceAccountOptions `json:"serviceAccount,omitempty"`
+
+  Deployment *ControllerDeployment `json:"deployment,omitempty"`
+}
+
+// PermissionsSpec defines the permissions that a package will require to
+// operate.
+type PermissionsSpec struct {
+  Rules []rbac.PolicyRule `json:"rules,omitempty"`
+}
+
+// PackageStatus represents the observed state of a Package.
+type PackageStatus struct {
+  runtimev1alpha1.ConditionedStatus `json:"conditionedStatus,omitempty"`
+  ControllerRef                     *corev1.ObjectReference `json:"controllerRef,omitempty"`
+}
+```
+
+The following changes will be made to the `Package` schema.
+
+Fields to be added:
+
+- `spec.infrastructureDefinitions`: list of `InfrastructureDefinition`s
+  installed by the package.
+- `spec.infrastructurePublications`: list of `InfrastructurePublication`s
+  installed by the package.
+- `spec.compositions`: list of `Composition`s installed by the package.
+- `spec.desiredState`: the desired state, which can be one of `Active` or
+  `Inactive`.
+- `spec.installJobRef`: a reference to the `Job` that was used to unpack the
+  package. The `PackageRevision` must be able to access it to create the objects
+  it installs.
+- `spec.revision`: the number revision that this `PackageRevision` is for its
+  parent `Package`. This is a strictly incrementing value set by the `Package`
+  controller. If an existing `PackageRevision` is transitioned from `Inactive`
+  to `Active`, its revision number will be set to one more than the greatest
+  revision number of all existing `PackageRevision`s for the `Package`. This
+  number is used by the `Package` controller to garbage collect old
+  `PackageRevision`s.
+- `spec.icon`: this replaces the `spec.icons` field and serves as a string
+  representing the SVG icon for a `PackageRevision`. See
+  [#1119](https://github.com/crossplane/crossplane/issues/1119) for more
+  information.
+
+Fields to be modified:
+
+- `spec.dependsOn`: this field is intended to be used for resolving
```

Comment:
Packages may require a minimum Crossplane version if they depend on new features that were added to an existing v1beta1 API (e.g. composition), which may be needed for additive composition enhancements like transform functions or support for additional config languages. Would it make sense to include an optional `minimumCrossplaneVersion` as part of `dependsOn` or elsewhere in the `Package` schema?